### ISSUE TITLE
RuleChain annotation equivalent

### ISSUE THREAD

**Issue by rmannibucau** [junit-team/junit4#1221]
Idea is to be able to chain rules but keeping track of each instance

One quick and simple but efficient solution is to add to @Rule an index/order method:

```
@Rule(index = 0)
private MyRule superRule = new MyRule(...);

@Rule(index = 1)
private MyRule2 superRule2 = new MyRule2(...);
```

 it allows to have the chain well formed automatically and the @Test to still use the rule instance to do things/access rule data.

**Comment by Tibor17** [comment]
This would be an elegant solution (default index=0) for ordering problem for produce methods/fields of rules and new proposal @rules on test method https://github.com/junit-team/junit/issues/1074#issuecomment-158158565.
WDYT?

**Comment by Tibor17** [comment]
btw. Nice to have the field private as in Roman's example.
I do not see any reason why JUnit requires public modifier on this field, maybe because they do not use setAccessible(true) which anyway would not harm the runtime of the test.

**Comment by kcooney** [comment]
We require methods and fields that are annotated with `@Test`, `@Rule` etc to be public because `setAccessible(true)` might not work if the tests run in an environment with a security manager installed. It would be unfortunate if tests passed in one environment (an IDE, for example) but not in another environment (your build tool).

One way to think about this: test methods and rules are functionality that are exposed by your test to JUnit, so they are an external API, so should be public.

**Comment by Tibor17** [comment]
@kcooney you'r right with security manager, thx.

**Comment by ctrimble** [comment]
It would be very cool if rules allowed you to specify the dependencies by name and let JUnit work out the correct order, sorting by name for unrelated rules.  For instance:

```
@Rule
MyRule1 rule1 = new MyRule1();

@Rule(before=["rule1"])
MyRule2 rule2 = new MyRule2();

@Rule(after=["rule2"])
MyRule3 rule3 = new MyRule3();
```

With the resulting execution order being:

```
rule2
rule1
rule3
```

**Comment by ctrimble** [comment]
I have this class in an old framework that can do the sorting https://github.com/ctrimble/xchain/blob/master/core/src/main/java/org/xchain/framework/util/DependencySorter.java .  That code is under the Apache License, but that could be worked out.

**Comment by stefanbirkner** [comment]
Fixed by #1445.
